_posts/2023-07-29-oauth-authentication-scala-http4s.md

```diff
@@ -0,0 +1,672 @@
+
+   ![step5](../images/oauth/step5.png)
+
+1. Click the update application button at the bottom of the page.
+
+### Configuring our application.
+
+In the following sections, we'll build our Scala application using [Http4s](https://blog.rockthejvm.com/http4s-tutorial/) for the routing and serving the application, Circe for JSON parsing, and Ciris for configuration.
+
+Create a resources folder inside your main folder. Then within it create an `apiConfig.json` file and add the Client Id and Client Secret we saved earlier from GitHub in the following format.
+
+```json
+{
+  "key": "0131cd2e9eacae1854c5",
+  "secret": "227............18c" // put your application secret
+}
+```
+
+Also, create a `serverConfig.json` file and add the following contents:
+
+```json
+{
+  "host": "localhost",
+  "port": "8080"
+}
+```
+
+Here we set the host to `localhost` and the port to `8080`.
+
+We will be using Ciris for configuration in the coming section. At this point, our directory structure should look like this:
+
+![directory](../images/oauth/dir1.png)
+
+Let's create an `ApiConfig.scala` file in the following path, `src/main/scala/com/xonal/config/ApiConfig.scala` and add the following code.
+
+```scala
+package com.xonal.config
+import ciris.Secret
+
+final case class ApiConfig(key: String, secret: Secret[String])
+```
+
+The `ApiConfig` final case class will hold the `key` and `secret` values when retrieved from `apiConfig.json`. Ciris provides a `Secret` class that replaces our `secret` value with the first 7 characters of the SHA-1 hash, convenient for passing around sensitive details within our application.
+
+When handling configuration with Ciris, values are passed around as `ConfigValue`s. Here's how the official ciris website defines a ConfigValue:
+
+> "`ConfigValue` is the central concept in the library. It represents a single configuration value or a composition of multiple values."
+
+Therefore we'll need to represent our values in this type before we start passing them around the application. Since we saved our values in `.json` files, we will need to decode the JSON strings, this is where `circe` and `ciris-circe` come into play.
+
+```scala
+package com.rockthejvm.config
+
+import java.nio.file.Paths
+import ciris.{ConfigDecoder,ConfigValue,file,Effect,Secret}
+import ciris.circe.circeConfigDecoder
+import io.circe.Decoder
+...
+
+object ApiConfig{
+  given apiDecoder: Decoder[ApiConfig] = Decoder.instance { h =>
+    for
+      key <- h.get[String]("key")
+      secret <- h.get[String]("secret")
+    yield ApiConfig(key, Secret(secret))
+  }
+
+  given apiConfigDecoder: ConfigDecoder[String, ApiConfig] =
+    circeConfigDecoder("ApiConfig")
+}
+
+val apiConfig: ConfigValue[Effect, ApiConfig] = file(
+  Paths.get("oauth/src/main/resources/apiConfig.json")
+).as[ApiConfig]
+```
+
+This section can be confusing so pay attention, we will start from the bottom up.
```

Comment:
"pay attention" sounds a bit imperative 😄 

"This section can be confusing, so we'll explain starting from the bottom."

Reply:
Updated